conda/misc.py

```diff
@@ -177,6 +159,48 @@ def explicit(
     txn.execute()
 
 
+def _get_package_record_from_specs(specs: list[str]) -> Iterable[PackageCacheRecord]:
+    """Given a list of specs, find the corresponding PackageCacheRecord. If
+    some PackageCacheRecords are missing, raise an error.
+    """
+    specs_pcrecs = tuple(
+        [spec, next(PackageCacheData.query_all(spec), None)] for spec in specs
+    )
+
+    # Assert that every spec has a PackageCacheRecord
+    specs_with_missing_pcrecs = [
+        str(spec) for spec, pcrec in specs_pcrecs if pcrec is None
+    ]
+    if specs_with_missing_pcrecs:
+        if len(specs_with_missing_pcrecs) == len(specs_pcrecs):
+            raise SpecNotFoundInPackageCache("No package cache records found")
+        else:
+            missing_precs_list = ", ".join(specs_with_missing_pcrecs)
+            raise SpecNotFoundInPackageCache(
+                f"Missing package cache records for: {missing_precs_list}"
+            )
+    return [rec[1] for rec in specs_pcrecs]
```

Comment:
Would `groupby_to_dict` help here?

```suggestion
    grouped_specs = groupby_to_dict(specs, lambda spec: next(PackageCacheData.query_all(spec), None) is not None)

    pcrecs_exists = grouped_specs.get(True) or []
    pcrecs_missing = grouped_specs.get(False) or []

    # Assert that every spec has a PackageCacheRecord
    if pcrecs_missing and not pcrecs_exists:
        raise SpecNotFoundInPackageCache("No package cache records found")
    elif specs_missing:
        missing_precs_list = ", ".join(pcrecs_missing)
        raise SpecNotFoundInPackageCache(
            f"Missing package cache records for: {missing_precs_list}"
        )
    return pcrecs_exists
```

Reply:
This doesn't quite work as written because conda needs to be returning the package cache records instead of the matchspecs. Made some slight modifications and it works. But might be worth a second look.

Reply:
See https://github.com/conda/conda/pull/14937#discussion_r2178605100 for followup